A couple of vulkan utilities

expipiplus1 opened the pull request:
###### Things done:
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] ~~OS X~~ (Not applicable)
  - [x] Linux
- [x] Tested execution of all binary files (usually in `./result/bin/`). See below
###### Notes

I don't think that this is quite ready to be merged as it's not easy to use the resultant binary or library.

In order to run `vulkaninfo` on Ubuntu it's necessary to do this at the moment: `LD_LIBRARY_PATH=/usr/lib/nvidia-355/:/usr/lib/x86_64-linux-gnu ~/.nix-profile/bin/vulkaninfo` 

I think that adding `/usr/lib/nvidia-355/` to `LD_LIBRARY_PATH` is unavoidable on Ubuntu due to #9415. Having to add `/usr/lib/x86_64-linux-gnu` is totally gross, and I'd rather avoid it if possible. This seems to be necessary to link with `libX11.so` and `libxcb-dri3.so`. Perhaps there should be some fancy library path patching in the mix.

I'd be grateful if someone on NixOS with a suitable driver installed could give `nix-shell -p vulkan-loader --command vulkaninfo` a spin and let me know what happens.

expipiplus1 committed:
glslang: init at 20160727

expipiplus1 committed:
vulkan-loader: init at 1.0.21.0

vcunat commented:
With 358.16 on NixOS I get
```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.7

Cannot create Vulkan instance.
/tmp/nix-build-vulkan-loader-1.0.7.0.drv-0/Vulkan-LoaderAndValidationLayers-cd99e567ae76ae6a6900b80b2c8f6031a194003a-src/demos/vulkaninfo.c:633: failed with VK_ERROR_INCOMPATIBLE_DRIVER
```

Apparently it's a too old driver.

expipiplus1 commented:
I get the correct behaviour with version 355.00.29 of the nvidia drivers.

Version 364.12 which claims to have vulkan support can be downloaded here https://developer.nvidia.com/vulkan-driver

I don't know what pains that will take to be installable on NixOS however.

vcunat commented:
I think the problem in my case might rather be incompatible GPU, even though the error message doesn't seem to fit much if that's so. I've got an older Fermi card.

grahamc commented:
I have a NVIDIA Corporation GK107M [GeForce GT 750M Mac Edition] and got the following output:

```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.7

Cannot create Vulkan instance.
/tmp/nix-build-vulkan-loader-1.0.7.0.drv-0/Vulkan-LoaderAndValidationLayers-cd99e567ae76ae6a6900b80b2c8f6031a194003a-src/demos/vulkaninfo.c:633: failed with VK_ERROR_INCOMPATIBLE_DRIVER

```

expipiplus1 commented:
It seems as though the 358 driver in NixOS is not vulkan ready. Is it possible to install 364.12?

grahamc commented:
We'd need to backport 59ef7d25c3944f2b46461181f6bc99f183389644 to release-16.03. @domenkozar  any opinions on backporting the nvidia driver upgrade?

expipiplus1 commented:
Now that the driver's been updated. Can there be any progress made on this?

grahamc commented:
When I rebased against master, I got:

```
[ 31%] Building CXX object glslang/CMakeFiles/glslang.dir/MachineIndependent/intermOut.cpp.o
/tmp/nix-build-glslang-2016-04-01.drv-0/glslang-a42533eca127aae1f5b4ee5e3fae357af0867ae1-src/glslang/MachineIndependent/intermOut.cpp: In function 'bool {anonymous}::is_positive_infinity(double)':
/tmp/nix-build-glslang-2016-04-01.drv-0/glslang-a42533eca127aae1f5b4ee5e3fae357af0867ae1-src/glslang/MachineIndependent/intermOut.cpp:52:10: error: '__builtin_isinf_sign' is not a member of 'std'
   return std::isinf(x) && (x >= 0);
          ^
/tmp/nix-build-glslang-2016-04-01.drv-0/glslang-a42533eca127aae1f5b4ee5e3fae357af0867ae1-src/glslang/MachineIndependent/intermOut.cpp:52:10: note: suggested alternative:
<built-in>: note:   '__builtin_isinf_sign'
make[2]: *** [glslang/CMakeFiles/glslang.dir/build.make:375: glslang/CMakeFiles/glslang.dir/MachineIndependent/intermOut.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:106: glslang/CMakeFiles/glslang.dir/all] Error 2
make: *** [Makefile:139: all] Error 2
builder for ‘/nix/store/s3znrjvaz7q6m1pbcfnxxzqm57pbflf3-glslang-2016-04-01.drv’ failed with exit code 2
error: build of ‘/nix/store/s3znrjvaz7q6m1pbcfnxxzqm57pbflf3-glslang-2016-04-01.drv’ failed
```

If I don't rebase, glslang and vulkan-info build okay... but:

```
$  ./result/bin/vulkaninfo 
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.7

Cannot create Vulkan instance.
/tmp/nix-build-vulkan-loader-1.0.7.0.drv-0/Vulkan-LoaderAndValidationLayers-cd99e567ae76ae6a6900b80b2c8f6031a194003a-src/demos/vulkaninfo.c:633: failed with VK_ERROR_INCOMPATIBLE_DRIVER
```

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nixos/nixpkgs/14404#-:-KMoe_xzUb1sUmI-LeBM:534357903)*
<!-- Sent from Reviewable.io -->

grahamc commented:
Do you plan on revisiting this PR?

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nixos/nixpkgs/14404#-:-KNdGrmcndNEw6PjYINr:546829910)*
<!-- Sent from Reviewable.io -->

expipiplus1 commented:
Sure, I'll try to take a look later today.

expipiplus1 commented:
@grahamc I've fixed glslang, vulkan-loader needs a little tweaking to get the latest version working still.

lukasoyen committed:
spirv-tools: init at 20160718

lukasoyen committed:
glslang: 2016-07-27 -> 2016-07-16

`vulkan-loader` requires a specific version of `glslang` as specified in
`<vulkan-loader-repo>/glslang_revision`.

lukasoyen committed:
vulkan-loader: add `spirv-tools` as dependency

lummax commented:
@expipiplus1 I took the liberty to update your changes and make it work. `vulkan-loader` needs a specific version  of `glslang` and `spirv-tools`, so I downgraded them. Feel free to incorporate these changes into this PR or let me know, if I should open a new one. https://github.com/lummax/nixpkgs/tree/vulkan-loader

cc @vcunat @grahamc

lukasoyen committed:
vulkan-loader: enable wayland WSI

expipiplus1 commented:
@lummax thanks! I've put those commits in this PR.

Ralith committed:
vulkan-loader: 1.0.21.0 -> 1.0.26.0

Also added asserts to ensure everything stays in sync.

Ralith committed:
shaderc-git: init at 2016-09-08

expipiplus1 commented:
@Ralith has added some more changes to the PR.

corngood commented:
I rebased this on master (#18863) to resolve the conflict.  Would one of you like to reset this branch to it?  Is there anything else I can do to help get this in?

expipiplus1 commented:
I've rebased on master (Thanks @corngood)

@grahamc This should be good to go now I think.

corngood commented on pkgs/development/libraries/vulkan-loader/default.nix:
```diff
@@ -0,0 +1,46 @@
+    rev = "sdk-${version}";
+    sha256 = "157m746hc76xrxd3qq0f44f5dy7pjbz8cx74ykqrlbc7rmpjpk58";
+  };
+  getRev = name: builtins.substring 0 40 (builtins.readFile "${src}/${name}_revision");
```
There's a CI error here

`error: cannot read ‘/nix/store/mll9kdy5pi9wirn088zpkh0hj2wk742y-Vulkan-LoaderAndValidationLayers-sdk-1.0.26.0-src/spirv-tools_revision’, since path ‘/nix/store/i5y4wagj1zjbd490ib4qls7zff6adpm2-Vulkan-LoaderAndValidationLayers-sdk-1.0.26.0-src.drv’ is not valid, at /home/travis/build/NixOS/nixpkgs/pkgs/development/libraries/vulkan-loader/default.nix:14:43`

vcunat replied:
This is no good, as it makes evaluation-time stuff (the assertions) depend on build-time stuff (the `src` download). We want to avoid that in nixpkgs, as the consequence is that nix would start building in unexpected situations, such as just making queries about packages. That's also why it fails on CI and it would fail on Hydra.

vcunat replied:
I converted the checks to run-time to resolve this: a307021

corngood commented:
I don't understand how that `getRev` is supposed to work.  Should we remove it?  

We could just do `preConfigure = "bash update_external_sources.sh";` instead.  Are the dependencies even needed for anything other than the demos?

corngood commented:
In case it's useful to anyone, I'm working on packaging AMDGPU-PRO, which includes a Vulkan icd: #18861

It works at least enough to run the sdk demos.

expipiplus1 commented:
@lummax ^

Ralith commented:
@corngood 
> I don't understand how that getRev is supposed to work. Should we remove it?

I'm not sure why it's failing, but it's important that we keep it or something very similar to ensure that revisions don't fall out of sync with what the package expects. What's unclear about its mode of operation? It reads the first 40 bytes from a file in the source which identifies a git hash.

> Are the dependencies even needed for anything other than the demos?

Yes, for example shaderc requires glslang.

corngood commented:
> I'm not sure why it's failing, but it's important that we keep it or something very similar to ensure that revisions don't fall out of sync with what the package expects. What's unclear about its mode of operation? It reads the first 40 bytes from a file in the source which identifies a git hash.

I don't know that there's something wrong with it.  I'm new to nix, and I couldn't find any other packages that reference files in the source directory like that.  Any idea why it's failing CI?

> Yes, for example shaderc requires glslang.

I was just thinking about vulkan-loader.  Using `bash update_external_sources.sh` works for building the loader and validation libraries, but without the demos it might not need the dependencies at all.

corngood commented:
I guess it's reading the file from the nix store without any guarantee that it exists?  If that's the case, I would either hard-code the dependency hashes or just remove the checks altogether.

Ralith commented:
Oh, right, evaluation can happen without the source being fetched. I've prepared a fix at https://github.com/expipiplus1/nixpkgs/pull/2.

corngood commented:
Yeah, that looks good.  Hopefully it'll pass CI now.  Can we take off the 'WIP'?

corngood commented:
I tried adding a libtool dependency to fix the osx build error, but it wasn't as simple as that (see my closed PR above).  Next error is

```
Scanning dependencies of target shaderc_combined_genfile
[ 92%] Generating libshaderc_combined.a
Usage: /nix/store/c5fzlwrfva8alj03b4r1haf670pi2mxy-libtool-2.4.6/bin/libtool [OPTION]... [MODE-ARG]...
Try 'libtool --help' for more information.
libtool:   error: unrecognised option: '-static'
```

I don't have access to an osx machine, and fixing it via travis is tedious.  Does anyone want to try to fix it?  Should we just disable it on osx?  It's at least theoretically useful with something like MoltenVK.

expipiplus1 commented:
Would that involve packaging up MoltenVK in nixpkgs?

It would probably be best to disable this build on OSX and get this merged for now.

corngood commented:
No, MoltenVK would just be another ICD.

Ralith commented:
I'm also in favor of disabling OSX support for now. It's probably fixable but there's no reason for it to block the rest until somebody with the right know-how shows up.

corngood commented:
* Allow i686 vulkan-loader (tested)
* Make shaderc linux-only

 #18940

It passes osx CI, but now fails on linux with a log that's truncated before anything useful.  Maybe just cherry-pick the shaderc change and see if that works?

Ralith commented:
@corngood It looks like one of two linux CI builds passed, and it certainly builds fine here. Do you know what the difference between them is?

corngood commented:
I'm not sure what the difference is.  It looks like it's deep in building a bunch of haskell stuff, which makes me think it's not pulling binary packages properly.

Perhaps I messed up the rebase, or the base revision doesn't have cached binaries?  Might be worth rebasing again on upstream latest.

corngood commented:
I have a few more changes on https://github.com/corngood/nixpkgs/commits/vulkan-loader to get the validation layers working.  It could be a separate PR.

@expipiplus1  are you comfortable removing WIP?

corngood commented:
@expipiplus1 What do you think about the name of `vulkan-loader` vs the `extra-libraries: vulkan` used in your haskell vulkan package.  The latter results in `librarySystemDepends = [ vulkan ];` in haskell-modules.nix, or if you use cabal2nix.  I suppose it could be aliased in all-packages.nix, or just renamed?

Ralith commented:
Outright naming the package `vulkan` seems reasonable, since it's the it's the non-hardware-specific component.

Ralith commented:
@corngood Is a module really the right way to make the layers visible? Since they're not used by default, I feel like they should just be directly installed in e.g. `/run/current-system/sw/share/vulkan/icd.d`, and the loader patched to use `XDG_DATA_DIRS` if it doesn't already.

corngood commented:
@Ralith Yeah, it really should be using /share/ instead of /etc/.  I also wanted to split the headers and layers into sub-packages.  I'll do some more work before I make a PR.

expipiplus1 commented:
It's certainly an inconvenience having the names clash like that. Changing the `vulkan` haskell package to reference `vulkan-loader` in `extra-libraries` might break things for non-nix builds. On the other hand, it prevents haskell-modules from working out of the box, that can be tweaked in configuration-general.nix though to reference `vulkan-loader` instead.

The name of this package isn't too important to me as long as it has an alias which isn't `vulkan`.

@corngood given that you seem to be driving this at the moment it might make sense for me to close this PR and for you to open one with your changes. Would this be OK?

corngood commented:
@expipiplus1 I don't mind doing a PR for this, but is there any problem with just merging it as is?  It just needs all-packages.nix to be resolved again.  What we've been discussing can be future work.

@vcunat What do you think about merging this as is?

expipiplus1 commented:
I've rebased on the latest master.

vcunat commented:
Picked to 16.09 release, too, as it's just addition of new packages.

Ralith commented:
@vcunat I'm confused about your patch a30702135d0f00acad24b94655636e963b24baf3 . That issue was fixed some time ago in https://github.com/expipiplus1/nixpkgs/pull/2 . Did that commit go missing from the PR somehow? Perhaps when @expipiplus1 rebased?

vcunat commented:
That's strange. In any case, now the commit isn't in this PR and I noticed because I tried just `nix-instantiate` first...
